contracts/badger-ren/BadgerRenAdapter.sol

```diff
@@ -0,0 +1,272 @@
+        percentageFeeRewardsBps = _feeConfig[2];
+        percentageFeeGovernanceBps = _feeConfig[3];
+
+        // Approve exchange.
+        require(renBTC.approve(_exchange, uint256(-1)));
+        require(wBTC.approve(_exchange, uint256(-1)));
+    }
+
+    function recoverStuck(
+        bytes calldata encoded,
+        uint256 _amount,
+        bytes32 _nHash,
+        bytes calldata _sig
+    ) external nonReentrant {
+        // Ensure sender matches sender of original tx.
+        uint256 start = encoded.length - 32;
+        address sender = abi.decode(encoded[start:], (address));
+        require(sender == msg.sender);
+
+        bytes32 pHash = keccak256(encoded);
+        uint256 _mintAmount = registry.getGatewayBySymbol("BTC").mint(pHash, _amount, _nHash, _sig);
+        uint256 _fee = _processFee(renBTC, _mintAmount, mintFeeBps);
+
+        emit RecoverStuckRenBTC(_mintAmount, _fee);
+
+        renBTC.safeTransfer(msg.sender, _mintAmount.sub(_fee));
+    }
+
+    function mintRenBTC(
+        // user args
+        address payable _destination,
+        // darknode args
+        uint256 _amount,
+        bytes32 _nHash,
+        bytes calldata _sig
+    ) external nonReentrant {
+        // Mint renBTC tokens
+        bytes32 pHash = keccak256(abi.encode(_destination));
+        uint256 _mintAmount = registry.getGatewayBySymbol("BTC").mint(pHash, _amount, _nHash, _sig);
+        uint256 _fee = _processFee(renBTC, _mintAmount, mintFeeBps);
+
+        emit MintRenBTC(_mintAmount, _fee);
+
+        renBTC.safeTransfer(_destination, _mintAmount.sub(_fee));
+    }
+
+    function burnRenBTC(bytes calldata _btcDestination, uint256 _amount) external nonReentrant {
+        require(renBTC.balanceOf(address(msg.sender)) >= _amount);
+        uint256 _startBalance = renBTC.balanceOf(address(this));
+        renBTC.safeTransferFrom(msg.sender, address(this), _amount);
+        uint256 _endBalance = renBTC.balanceOf(address(this));
+
+        uint256 _burnAmount = _endBalance.sub(_startBalance);
+        uint256 _fee = _processFee(renBTC, _burnAmount, burnFeeBps);
+
+        emit BurnRenBTC(_burnAmount, _fee);
+
+        uint256 burnAmount = registry.getGatewayBySymbol("BTC").burn(_btcDestination, _burnAmount.sub(_fee));
+    }
+
+    function mintWBTC(
+        uint256 _slippage,
+        address payable _destination,
+        uint256 _amount,
+        bytes32 _nHash,
+        bytes calldata _sig
+    ) external nonReentrant {
+        // Mint renBTC tokens
+        bytes32 pHash = keccak256(abi.encode(_slippage, _destination));
+        uint256 _mintAmount = registry.getGatewayBySymbol("BTC").mint(pHash, _amount, _nHash, _sig);
+        _swapRenBTCForWBTC(_slippage, _destination, _mintAmount);
+    }
+
+    function _swapRenBTCForWBTC(
+        uint256 _slippage,
+        address payable _destination,
+        uint256 _mintAmount
+    ) internal {
+        // Get price
+        uint256 dy = exchange.get_dy(0, 1, _mintAmount);
```

Comment:
Given we rely 100% on curve.fi to give us the best price(`get_dy`), if pool is unbalanced or swap amount is large it can lead to unfavourable x rate.  To combat this should we consider interrogating oracle for the best price or alternatively consider https://curve.readthedocs.io/registry-exchanges.html#Swaps.get_best_rate? 

Should we consider augmenting user experience with displaying estimated value for the mint operation that involves the swap?